Studio: Add Song of the Day functionality

panpawn opened the pull request:
Requested by #Don't Lose

Slayer95 commented on chat-plugins/thestudio.js:
```diff
@@ -257,11 +257,35 @@ let commands = {
+		}
+
+		let title = target.split(',')[0].trim();
+		let link = target.split(',')[1].trim();
```
This crashes if the target has no comma

Slayer95 commented on chat-plugins/thestudio.js:
```diff
@@ -257,11 +257,35 @@ let commands = {
+		let title = target.split(',')[0].trim();
+		let link = target.split(',')[1].trim();
+
+		room.chatRoomData.songOfTheDay = `<a href="${link}" title="${title}">${Chat.escapeHTML(title)}</a>.`;
```
The attributes need HTML escaping

Slayer95 commented on chat-plugins/thestudio.js:
```diff
@@ -257,11 +257,35 @@ let commands = {
+	},
+	songhelp: [
+		"/aotd song [song], [link] - Sets the Song of the Day. Requires % @ * # & ~",
+		"/oatd song [off|disable] - Turns the Song of the Day off. Requires 5 @ * # & ~",
```
Tipo

> Requires 5 @ * # &

Slayer95 commented on chat-plugins/thestudio.js:
```diff
@@ -285,5 +309,7 @@ exports.commands = {
 		"- /aotd quote - View the current Artist Quote of the Day.",
 		"- /aotd quote [quote] - Set the Artist Quote of the Day. Requires: # & ~",
+		"- /aotd song [song], [link] - Sets the Song of the Day. Requires % @ * # & ~",
+		"- /oatd song [off|disable] - Turns the Song of the Day off. Requires 5 @ * # & ~",
```
Typo in the namespace

Slayer95 replied:
Also in the permissions

panpawn commented:
It's good to have you back @Slayer95 :o)

I believe I've updated this with all of the requested changes thus far, I must of been really tired when I made this originally...

Slayer95 commented on chat-plugins/thestudio.js:
```diff
@@ -257,11 +257,37 @@ let commands = {
+			let title = target.split(',')[0].trim();
+			let link = target.split(',')[1].trim();
+
+			room.chatRoomData.songOfTheDay = `<a href="${link}" title="${Chat.escapeHTML(title)}">${Chat.escapeHTML(title)}</a>.`;
```
The ``href`` *still* needs escaping...

Slayer95 replied:
``Chat.html`` should do the job here, actually.

Slayer95 commented on chat-plugins/thestudio.js:
```diff
@@ -257,11 +257,37 @@ let commands = {
 		if (!room.chatRoomData || !this.runBroadcast()) return false;
-		this.sendReplyBox("The Artist of the Day " + (room.chatRoomData.artistOfTheDay ? "is " + room.chatRoomData.artistOfTheDay + "." : "has not been set yet."));
+		let song = (room.chatRoomData.songOfTheDay ? '<br />The Song of the Day is ' + room.chatRoomData.songOfTheDay : '');
+		this.sendReplyBox("The Artist of the Day " + (room.chatRoomData.artistOfTheDay ? "is " + room.chatRoomData.artistOfTheDay + "." : "has not been set yet.") + song);
```
While you are at this, it seems that ``artistOfTheDay`` may contain special chars, but I may be wrong

panpawn replied:
No, you're right! Fixed :p

panpawn commented:
Updated @Slayer95 

You were right, there was a lack of escaping with the AotD.

I also updated this to use ``Chat#html`` for where I use that template string.

Slayer95 commented:
I'm not a fan of storing escaped strings / markup, but I'll let someone else champion that.

Zarel commented:
I agree, escape on display, not on storage.

panpawn commented:
Done

Zarel commented on chat-plugins/thestudio.js:
```diff
@@ -257,11 +257,37 @@ let commands = {
+			let title = target.split(',')[0].trim();
+			let link = target.split(',')[1].trim();
+
+			room.chatRoomData.songOfTheDay = Chat.html`<a href="${link}" title="${title}">${title}</a>.`;
```
This should probably be called `songOfTheDayHTML` if it contains HTML, as it does.

panpawn commented:
done

Zarel commented on chat-plugins/thestudio.js:
```diff
@@ -257,11 +257,37 @@ let commands = {
 		if (!room.chatRoomData || !this.runBroadcast()) return false;
-		this.sendReplyBox("The Artist of the Day " + (room.chatRoomData.artistOfTheDay ? "is " + room.chatRoomData.artistOfTheDay + "." : "has not been set yet."));
+		let song = (room.chatRoomData.songOfTheDayHTML ? '<br />The Song of the Day is ' + room.chatRoomData.songOfTheDayHTML : '');
+		this.sendReplyBox("The Artist of the Day " + (room.chatRoomData.artistOfTheDay ? "is " + room.chatRoomData.artistOfTheDay + "." : "has not been set yet.") + song);
```
This does not appear to correctly escape the artist of the day.

panpawn replied:
ops

panpawn commented:
Updated
